source/extensions/filters/udp/udp_proxy/udp_proxy_filter.cc

```diff
@@ -37,18 +34,26 @@ void UdpProxyFilter::onData(Network::UdpRecvData& data) {
   active_session->write(*data.buffer_);
 }
 
+void UdpProxyFilter::onReceiveError(Api::IoError::IoErrorCode) {
+  config_->stats().downstream_sess_rx_errors_.inc();
+}
+
 UdpProxyFilter::ActiveSession::ActiveSession(UdpProxyFilter& parent,
                                              Network::UdpRecvData::LocalPeerAddresses&& addresses,
                                              const Upstream::HostConstSharedPtr& host)
     : parent_(parent), addresses_(std::move(addresses)), host_(host),
+      idle_timer_(parent.read_callbacks_->udpListener().dispatcher().createTimer(
+          [this] { onIdleTimer(); })),
       // NOTE: The socket call can only fail due to memory/fd exhaustion. No local ephemeral port
       //       is bound until the first packet is sent to the upstream host.
-      io_handle_(host->address()->socket(Network::Address::SocketType::Datagram)),
+      io_handle_(parent.createIoHandle(host)),
```

Comment:
> Is this change just for testing? It's a bit weird to add this indirection. Would it be better to just use MockOsSysCall? So there is no need to mock IoHandle. IoHandle is just a thin wrapper of OsSysCall, so far we haven't have the need to mock this layer.

Yes it's for testing. For sendmsg/recvmsg, it's actually not a thin wrapper, it does quite a lot with control messages. I would really prefer to not have to satisfy that code with a MockOsSysCall. IMO this is much simpler, so I would strongly prefer to keep it this way.

> I also think about changing the IoHandle here to ClientSocket so it's more clear when we use it in write()? But this is just a code optimization and has nothing to do with this PR. Your call.

Sorry can you expand?

Reply:
> > Is this change just for testing? It's a bit weird to add this indirection. Would it be better to just use MockOsSysCall? So there is no need to mock IoHandle. IoHandle is just a thin wrapper of OsSysCall, so far we haven't have the need to mock this layer.
> 
> Yes it's for testing. For sendmsg/recvmsg, it's actually not a thin wrapper, it does quite a lot with control messages. I would really prefer to not have to satisfy that code with a MockOsSysCall. IMO this is much simpler, so I would strongly prefer to keep it this way.

Can we add an indirect constructor which takes a IoHandlePtr for test purpose? Current constructor can call host->address()->socket() to create an IoHandle instance and pass it down to the indirect constructor.  In the unit test, you can use the indirect constructor with a MockIoHandle. In this way you won't need createIoHandle() interface for this class.

> > I also think about changing the IoHandle here to ClientSocket so it's more clear when we use it in write()? But this is just a code optimization and has nothing to do with this PR. Your call.
> 
> Sorry can you expand?

This io_handle, together with host_.address(), is used to write to backend service. I'm thinking if we should change it to a connection socket which can store up both io_handle and peer address. We need host_ in this class just to access its address() as of now. If this is true for the future, can we pack it in the connection socket together with the passed-in io_handle? But again, this is just a code optimization, we don't have to do it now.

Reply:
> Can we add an indirect constructor which takes a IoHandlePtr for test purpose? Current constructor can call host->address()->socket() to create an IoHandle instance and pass it down to the indirect constructor. In the unit test, you can use the indirect constructor with a MockIoHandle. In this way you won't need createIoHandle() interface for this class.

I will chat with out on Slack as I don't completely follow. There is an IOHandle per session, so I'm not sure how a different constructor would help? I think we need a factory.

> This io_handle, together with host_.address(), is used to write to backend service. I'm thinking if we should change it to a connection socket which can store up both io_handle and peer address. We need host_ in this class just to access its address() as of now. If this is true for the future, can we pack it in the connection socket together with the passed-in io_handle? But again, this is just a code optimization, we don't have to do it now.

Yes I agree I would like to have some kind of `ClientSocket` concept, but I looked into that and it would have not been a small change. I will add a TODO.

Reply:
> > Can we add an indirect constructor which takes a IoHandlePtr for test purpose? Current constructor can call host->address()->socket() to create an IoHandle instance and pass it down to the indirect constructor. In the unit test, you can use the indirect constructor with a MockIoHandle. In this way you won't need createIoHandle() interface for this class.
> 
> I will chat with out on Slack as I don't completely follow. There is an IOHandle per session, so I'm not sure how a different constructor would help? I think we need a factory.

Sorry, I missed the fact that test doesn't control how ActiveSession is created. I suggested offline that we can mock out the whole MockHost chain to achieve that which might needs extra work. But I'm okay with current solution as well.

Reply:
I will keep the current solution as that is a pattern we do in other places as well and it seems fine to me.